File: examples/typescript-api/src/domain/user/entity.ts
/**
 * User Entity
 *
 * 사용자 도메인 엔터티입니다.
 * 비즈니스 규칙과 불변성을 캡슐화합니다.
 */

import { Email } from './value-objects';

export class User {
  private constructor(
    public readonly id: string,
    public readonly email: Email,
    public readonly name: string,
    private _status: 'pending' | 'active' | 'inactive'
  ) {}

  static create(data: { email: string; name: string }): User {
    const email = Email.create(data.email);
    return new User(
      crypto.randomUUID(),
      email,
      data.name,
      'pending'
    );
  }

  static reconstitute(data: {
    id: string; email: string; name: string; status: string
  }): User {
    return new User(
      data.id,
      Email.create(data.email),
      data.name,
      data.status as 'pending' | 'active' | 'inactive'
    );
  }

  get status() { return this._status; }

  activate(): void {
    if (this._status !== 'pending') {
      throw new Error('Only pending users can be activated');
    }
    this._status = 'active';
  }

  deactivate(): void {
    if (this._status !== 'active') {
      throw new Error('Only active users can be deactivated');
    }
    this._status = 'inactive';
  }
}

File: examples/python-api/app/domain/user/entity.py
"""User Entity - Domain model."""

from dataclasses import dataclass
import re
import uuid


@dataclass
class User:
    id: str
    email: str
    name: str
    status: str = "pending"

    @classmethod
    def create(cls, email: str, name: str) -> "User":
        if not re.match(r"^[^\s@]+@[^\s@]+\.[^\s@]+$", email):
            raise ValueError(f"Invalid email: {email}")
        return cls(
            id=str(uuid.uuid4()),
            email=email.lower(),
            name=name,
            status="pending",
        )

    def activate(self) -> None:
        if self.status != "pending":
            raise ValueError("Only pending users can be activated")
        self.status = "active"

    def deactivate(self) -> None:
        if self.status != "active":
            raise ValueError("Only active users can be deactivated")
        self.status = "inactive"
